### PR TITLE
Add MacOSXWatchService

### PR DESCRIPTION
**Pull request by eatkins**
This watch service is taken from my CloseWatch plugin, found at
https://github.com/swoval/swoval

It uses a native jni library, provided by apple-file-events, to receive
file events from the os. I more or less copied this file from the close
watch source code with some minor modifications given that it is now in
the sbt.io package.

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.1.x&zinc=1.1.x&io=pull/138/head&librarymanagement=1.1.x&util=1.1.x&website=1.1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

File: io/src/main/scala/sbt/io/MacOSXWatchService.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+
+class MacOSXWatchService(watchLatency: Duration, val queueSize: Int = 256) extends WatchService {
+
+  private[this] val executor =
+    Executors.newSingleThreadExecutor(new ThreadFactory("sbt.io.MacOSXWatchService"))
+  private[this] val watcher: FileEventsApi = FileEventsApi.apply(onFileEvent, _ => {})
```
Comment (eed3si9n) [inline io/src/main/scala/sbt/io/MacOSXWatchService.scala]: This failed compilation on Travis for 2.10:

```scala
[error] /home/travis/build/sbt/io/io/src/main/scala/sbt/io/MacOSXWatchService.scala:21:66: missing arguments for method onFileEvent in class MacOSXWatchService;
[error] follow this method with `_' if you want to treat it as a partially applied function
[error]   private[this] val watcher: FileEventsApi = FileEventsApi.apply(onFileEvent, _ => {})
[error]                                                                  ^
```

File: io/src/main/scala/sbt/io/MacOSXWatchService.scala [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+          registered += path -> key
+          val flags = new Flags.Create().setNoDefer().setFileEvents().value
+          watcher.createStream(path.toRealPath().toString, watchLatency.toMillis / 1000.0, flags)
+          key
+      }
+    }
+  }
+
+  def onFileEvent(fileEvent: FileEvent): Unit = {
+    executor.submit((() => {
```
Comment (eed3si9n) [inline io/src/main/scala/sbt/io/MacOSXWatchService.scala]:
```scala
[error] /home/travis/build/sbt/io/io/src/main/scala/sbt/io/MacOSXWatchService.scala:61:25: type mismatch;
[error]  found   : () => Unit
[error]  required: Runnable
[error]     executor.submit((() => {
[error]                         ^
```

**Review by eed3si9n — CHANGES_REQUESTED** [review]
Travis CI failed for Scala 2.10. Probably need to make small adjustments to supply type annotation or sugar.

**Comment by eatkins** [comment]
I made the file 2.10 compatible, but, out of curiosity, why do we build for 2.10 (and 2.11)? Isn't this module only used by sbt 1.x?

**Comment by eed3si9n** [comment]
> I made the file 2.10 compatible, but, out of curiosity, why do we build for 2.10 (and 2.11)? Isn't this module only used by sbt 1.x?

We first split out IO as an aspirational general-purpose IO module while we were still in 0.13, and built other things on top of that. At some point we might drop 2.10.

**Review by eed3si9n — APPROVED** [review]

**Review by dwijnand — APPROVED** [review]
thanks @eatkins

**Pull request merged**